### DISCUSSION TITLE
Add restrictions on a docker workspace

### DISCUSSION THREAD

**Discussion by Y3NK** [coder/coder#15913] [General] [answered]
Hello, 
I want to add limitation on usage; CPU/DDR/GPU

nvidia container toolkit has been added.

My idea was to use:

```
data "coder_parameter" "memory_limit"  {
  name        = "memory_limit"
  description = "Gb of DDR ?"
  # icon        = "/emojis/1f30f.png"
  type        = "string"
  default     = "4g"

  option {
    name = "4GB"
    value = "4g"
    # icon = "/emojis/1f1fa-1f1f8.png"
  }

  option {
    name = "8GB"
    value = "8g"
    # icon = "/emojis/1f1fa-1f1f8.png"
  }

  option {
    name = "32GB"
    value = "32g"
    # icon = "/emojis/1f1fa-1f1f8.png"
  }
}
```

and 
  ```
command = [
    # "-e", "MEMORY=${data.coder_parameter.docker_host.value}g"
    "-m", "${data.coder_parameter.memory_limit.value}"
  ]
```

but no limits are applied...

Thanks @matifali  but where should I use the 
[gpus](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container#gpus-1) attribute and install the [nvidia-container-toolkit](https://github.com/NVIDIA/nvidia-container-toolkit).
?
I wanted to use either MIG limitation or gpu x, not all.

Or is it a best approach to look for integrate a docker-compose instead of docker image to use limits with something like :
```
deploy:
  resources:
    limits:
      cpus: '0.001'
      memory: 50M
    reservations:
      cpus: '0.0001'
      memory: 20M
```
?

Thanks

**Comment by matifali** [comment, marked as answer]
Please take a look at `cpu`, `memeory` and `gpu` arguments for [`docker_container`](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container) resource. You can play wit them to set limits as desired on the workspace container.

**Reply by Y3NK** [reply]
Thank you,
So these values should directly be used as is on main.tf ?

Have you any example of how to use docker-compose inside of Coder ? Because typically `gpu `usage from terraform only allow "all", where I would prefer to use other value (0,1 or else).

I am using Docker behind.

Thanks for your help

**Reply by matifali** [reply]
> So these values should directly be used as is on main.tf ?

Yes, you can also make use of `coder_parameter` to ask the user to choose or enter them.

>  Because typically gpu usage from terraform only allow "all", where I would prefer to use other value (0,1 or else).

Coder only supports Terraform as a backend. 

>  Because typically gpu usage from terraform only allow "all", where I would prefer to use other value (0,1 or else).

It's unfortunate that the docker provider does not support individual GPUs. Maybe create an issue to support this on https://github.com/kreuzwerker/terraform-provider-docker/issues

**Comment by Y3NK** [comment]
Thank you.
But passing an environment variable is possible, isn't ?
Meaning that passing -e NVIDIA_VISIBLE_DEVICE=0 should work, no ? same with NVIDIA_DRIVER_CAPABILITIES or NVIDIA_MIG_CONFIG_DEVICES ?
That would be the best option ?

**Reply by matifali** [reply]
You can test. I have not tried this yet.
